### PR TITLE
IGNITE-13191 Public-facing API for waiting for backups on shutdown

### PR DESCRIPTION
**Pull request by vldpyatkov**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java [hunk 1]
```diff
@@ -2623,20 +2657,44 @@ private static CacheConfiguration utilitySystemCache() {
...
+         */
+        private ShutdownPolicy determineShutdownPolicy() {
+            if (IgniteSystemProperties.getBoolean(IgniteSystemProperties.IGNITE_WAIT_FOR_BACKUPS_ON_SHUTDOWN))
+                return ShutdownPolicy.GRACEFUL;
+
+            return grid.cluster().shutdownPolicy();
+        }
+
+        /**
+         * Stop instance synchronously according to parameters.
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java:2691]: Stop**s**

File: modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java [hunk 2]
```diff
@@ -2623,20 +2657,44 @@ private static CacheConfiguration utilitySystemCache() {
...
+         *
+         * @param cancel Flag indicating whether all currently running job
          *      should be cancelled.
+         * @param shutdown Policy of according which shutdown will be produced.
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java:2695]: Policy according to which shutdown will be performed.

File: modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java [hunk 3]
```diff
@@ -2661,6 +2719,131 @@ private synchronized void stop0(boolean cancel) {
...
+
+                        if (fullMap == null) {
+                            safeToStop = false;
+
+                            break;
+                        }
+
+                        nodesToExclude.retainAll(fullMap.keySet());
+
+                        if (!haveCopyLoclaPartitions(grpCtx, nodesToExclude, proposedSuppliers)) {
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java:2784]: Typo: local

File: modules/core/src/main/java/org/apache/ignite/ShutdownPolicy.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite;
+
+public enum ShutdownPolicy {
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/ShutdownPolicy.java:20]: Let's describe shortly in class header what this property is responsible for.

File: modules/core/src/main/java/org/apache/ignite/ShutdownPolicy.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    /**
+     * Stop immediately as soon as all components are ready.
+     */
+    IMMEDIATE(0),
+
+    /**
+     * Node will stop if and only if it does not store any unique partitions, that does not have copies on cluster.
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/ShutdownPolicy.java:27]: , that don't have another copies in the cluster.

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/shutdown/ShutdownPolicyArgument.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.commandline.shutdown;
+
+import org.apache.ignite.ShutdownPolicy;
+
+/**
+ * Srgumants of shutdown policy command.
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/shutdown/ShutdownPolicyArgument.java:23]: Arguments ?

File: modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java [hunk 7]
```diff
@@ -2694,6 +2877,85 @@ else if (grid0.context().invalid())
...
+                    continue;
+
+                boolean foundCopy = false;
+
+                for (Map.Entry<UUID, GridDhtPartitionMap> entry : fullMap.entrySet()) {
+                    if (localNodeId.equals(entry.getKey()) || nodesToExclude.contains(entry.getKey()))
+                        continue;
+
+                    //This remote node does not present in ideal assignment.
+                    if (!idealAssignment.get(p).stream().filter(node -> node.id().equals(entry.getKey())).findAny().isPresent())
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java:2915]: image

File: modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java [hunk 8]
```diff
@@ -2694,6 +2877,85 @@ else if (grid0.context().invalid())
             }
         }
 
+        /**
+         * @param grpCtx Cahce group.
+         * @param nodesToExclude Nodes to exclude from check.
+         * @param proposedSuppliers Map of proposed suppliers for groups.
+         * @return True if all local partition of group specified have a copy in cluster, false otherwise.
+         */
+        private boolean haveCopyLoclaPartitions(
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java:2886]: This method implicitly fills proposedSuppliers map, let's describe it in the doc.

File: modules/core/src/main/java/org/apache/ignite/internal/CheckCpHistTask.java [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+import org.apache.ignite.internal.processors.task.GridInternal;
+import org.apache.ignite.resources.IgniteInstanceResource;
+import org.apache.ignite.resources.LoggerResource;
+
+/**
+ * Task for test last checkpoint about applicable of history to all groups and partitions of parameters.
+ * If one of group or partition does not applicable force checkpoint will trigger.
+ */
+@GridInternal
+public class CheckCpHistTask extends ComputeTaskAdapter<Map<UUID, Map<Integer, Set<Integer>>>, Boolean> {
```
[the remark this answers is not in the document]
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/CheckCpHistTask.java:47]: Look, I did it.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/configuration/distributed/DistributedEnumProperty.java [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+
+import java.io.Serializable;
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.internal.util.future.GridFutureAdapter;
+import org.apache.ignite.internal.util.typedef.internal.S;
+import org.apache.ignite.lang.IgniteClosure;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Wrapper for any enumerator.
```
[the remark this answers is not in the document]
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/configuration/distributed/DistributedEnumProperty.java:28]: I corrected this description.

File: modules/core/src/main/java/org/apache/ignite/IgniteCluster.java [hunk 11]
```diff
@@ -629,4 +629,19 @@
...
+    /**
+     * Return a policy of shutdown or default walue {@code IgniteConfiguration.DFLT_SHUTDOWN_POLICY}
+     * if property is not set.
+     *
+     * @return Shutdown policy.
+     */
+    public ShutdownPolicy shutdownPolicy();
+
+    /**
+     * Set a shutdown policy on a cluster.
```
[the remark this answers is not in the document]
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/IgniteCluster.java:642]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java (hunk 8 above)
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java:2886]: Added method description.

File: modules/core/src/main/java/org/apache/ignite/ShutdownPolicy.java (hunk 4 above)
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/ShutdownPolicy.java:20]: Done.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/dht/GracefulShutdownTest.java [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+import org.junit.Test;
+
+import static org.apache.ignite.IgniteSystemProperties.IGNITE_WAIT_FOR_BACKUPS_ON_SHUTDOWN;
+import static org.apache.ignite.internal.IgniteNodeAttributes.ATTR_IGNITE_INSTANCE_NAME;
+
+/**
+ * Test checks various cluster shutdown and initiated policy.
+ */
+@WithSystemProperty(key = IGNITE_WAIT_FOR_BACKUPS_ON_SHUTDOWN, value = "false")
+public class GracefulShutdownTest extends GridCacheDhtPreloadWaitForBackupsWithPersistenceTest {
```
[the remark this answers is not in the document]
Comment (vldpyatkov) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/dht/GracefulShutdownTest.java:47]: I see each test has one line description.
If you think about full scenario in test description I can add it. But I suppose that does not easier for understanding than read a code.

File: modules/core/src/main/java/org/apache/ignite/ShutdownPolicy.java [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite;
+
+/**
+ * This class reperesents a policy of shutdown.
+ * The policy gives a guaranties data safety on ordinal shutdown one or more cluster nodes.
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/ShutdownPolicy.java:22]: Did you mean "ordinary"?
The policy specifies data safety guarantees on ordinary shutdown of one or more cluster nodes.

File: modules/core/src/test/java/org/apache/ignite/internal/commandline/CommandHandlerParsingTest.java [hunk 14]
```diff
@@ -330,6 +333,26 @@ public void testParseAndValidateWalActions() {
         assertParseArgsThrows("Unexpected action " + rnd + " for " + WAL.text(), WAL.text(), rnd);
     }
 
+    /**
+     * Tets checks a parser of shutdown policy command.
+     */
+    @Test
+    public void tesParseShutdownPolicyParameters() {
```
Comment (glukos) [inline modules/core/src/test/java/org/apache/ignite/internal/commandline/CommandHandlerParsingTest.java:340]: Typo: test

File: modules/core/src/test/java/org/apache/ignite/internal/commandline/CommandHandlerParsingTest.java [hunk 15]
```diff
@@ -337,7 +337,7 @@ public void testParseAndValidateWalActions() {
      * Tets checks a parser of shutdown policy command.
      */
     @Test
-    public void tesParsShutdownPolicyParameters() {
+    public void tesParseShutdownPolicyParameters() {
```
Comment (glukos) [inline modules/core/src/test/java/org/apache/ignite/internal/commandline/CommandHandlerParsingTest.java:340]: typo: test

File: modules/core/src/main/java/org/apache/ignite/ShutdownPolicy.java (hunk 13 above)
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/ShutdownPolicy.java:22]: Done.

File: modules/core/src/test/java/org/apache/ignite/internal/commandline/CommandHandlerParsingTest.java (hunk 14 above)
Comment (vldpyatkov) [reply modules/core/src/test/java/org/apache/ignite/internal/commandline/CommandHandlerParsingTest.java]: Fixed.

File: modules/core/src/test/java/org/apache/ignite/internal/commandline/CommandHandlerParsingTest.java (hunk 15 above)
Comment (vldpyatkov) [reply modules/core/src/test/java/org/apache/ignite/internal/commandline/CommandHandlerParsingTest.java]: Same as above.

**Pull request closed without merging**